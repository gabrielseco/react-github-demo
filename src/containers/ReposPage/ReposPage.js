// @flow
import React, { Component } from 'react';
import debounce from 'lodash.debounce';
import { Spinner, ReposList } from './../../components';
import GithubApi from './../../services/GithubApi';
import styles from './ReposPage.scss';
import { GithubRepo } from './../../types';

type State = {
  repos: GithubRepo[],
  isLoading: boolean
};

type Props = {
  username: string
};

class ReposPage extends Component<Props, State> {
  constructor(props: any) {
    super(props);
    this.state = {
      repos: [],
      isLoading: this.props.isLoading
    };

    this.searchRepos = debounce(this.searchRepos, 500);
  }

  componentWillReceiveProps(props) {
    this.setState(prevState => {
      return {
        ...prevState,
        isLoading: props.isLoading
      };
    });

    if (props.username.trim() !== '') {
      this.searchRepos(props.username);
    } else {
      this.setState(prevState => {
        return {
          ...prevState,
          repos: [],
          isLoading: false
        };
      });
    }
  }

  async searchRepos(username: string) {
    if (username !== '') {
      const repos = await GithubApi.getRepos(username);

      this.setState(prevState => {
        return {
          ...prevState,
          repos: repos,
          isLoading: false
        };
      });
    }
  }

  renderResults(repos: GithubRepo[]) {
    if (repos.length === 0) {
      return null;
    }
    return (
      <div className={styles.reposContainer}>
        <ReposList repos={repos} />
      </div>
    );
  }

  render() {
    const { repos, isLoading } = this.state;
    return (
      <div>
        {isLoading ? (
          <div className={styles.spinnerContainer}>
            <Spinner />
          </div>
        ) : (
          this.renderResults(repos)
        )}
      </div>
    );
  }
}

export default ReposPage;
