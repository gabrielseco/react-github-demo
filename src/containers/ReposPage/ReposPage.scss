.spinnerContainer {
  display: flex;
  justify-content: center;
  margin-top: 50px;
}

.reposContainer {
  padding: 20px;
}
