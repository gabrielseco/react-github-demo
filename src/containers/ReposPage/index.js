export { default as ReposPage } from './ReposPage';
