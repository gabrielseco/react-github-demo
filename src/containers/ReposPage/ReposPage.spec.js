import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import ReposPage from './ReposPage';

describe('ReposPage suite', () => {
  it('renders RepoPage', () => {
    const comp = <ReposPage isLoading username="gabrielseco" />;
    const component = shallow(comp);
    const tree = renderer.create(comp).toJSON();
    expect(component).toBeDefined();
    expect(tree).toMatchSnapshot();
  });
});
