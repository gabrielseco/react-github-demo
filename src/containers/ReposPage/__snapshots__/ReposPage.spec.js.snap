// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReposPage suite renders RepoPage 1`] = `
<div>
  <div
    className="spinnerContainer"
  >
    <div
      aria-busy="true"
      aria-live="assertive"
      className="spinner"
      role="alertdialog"
    />
  </div>
</div>
`;
